### PR TITLE
Wasmvm upgrade 557

### LINKED ISSUE #557: Update to wasmvm / cosmwasm 0.16
There will be one more update before

Key Changes:
- Lots of changes to the `IbcChannel*Msg` types
- CounterPartyVersion is no longer in `IbcChannel` but in some of those messages
- Change standard `contract_address` event attribute (for `wasm` type) to `_contract_address`.
- Enforce the contract cannot return any `_*` attributes and thus conflict with any special "trusted" attributes from the runtime.
- Handle custom events from `IbcBasicResponse` and `IbcReceiveResponse` (not just attributes)

### PR DESCRIPTION
**Pull request by ethanfrey**
Closes #557 

Currently tested versions:

- Contracts: cosmwasm v0.16.-rc4
- Engine: wasmvm v0.16.0-rc2

Work done:

- [x] Update wasmd and test contracts 
- [x] Lots of changes to the `IbcChannel*Msg` types
- [x] CounterPartyVersion is no longer in `IbcChannel` but in some of those messages
- [x] Change standard `contract_address` event attribute (for `wasm` type) to `_contract_address`.
- [x] Enforce the contract cannot return any `_*` attributes and thus conflict with any special "trusted" attributes from the runtime.
- [x] Handle custom events from `IbcBasicResponse` and `IbcReceiveResponse` (not just attributes)

### CODE REVIEW

File: x/wasm/keeper/events.go [hunk 1]
```diff
@@ -50,8 +51,9 @@ func contractSDKEventAttributes(customAttributes []wasmvmtypes.EventAttribute, c
 	attrs := []sdk.Attribute{sdk.NewAttribute(types.AttributeKeyContractAddr, contractAddr.String())}
 	// append attributes from wasm to the sdk.Event
 	for _, l := range customAttributes {
-		// and reserve the contract_address key for our use (not contract)
-		if l.Key != types.AttributeKeyContractAddr {
+		// FIXME: do we want to error here on invalid events
+		// and reserve all _* keys for our use (not contract)
```
Comment (webmaster128) [inline x/wasm/keeper/events.go:55]: We do!

Comment (webmaster128) [reply x/wasm/keeper/events.go:55]: Same for the min length check in `newCustomEvents` above. Why silently ignore events with invalid type length?

Comment (ethanfrey) [reply x/wasm/keeper/events.go:55]: This was Alex's choice.
I left a message on the v0.15.0 upgrade PR to that effect. (See https://github.com/CosmWasm/wasmd/pull/548#discussion_r674203149 )

I would make that change in a separate PR as it is unrelated to v0.16.0. But good not to miss it.

Comment (webmaster128) [reply x/wasm/keeper/events.go:55]: Could you create a ticket for those two places? Then we can merge here.

File: x/wasm/keeper/relay.go [hunk 2]
```diff
@@ -60,8 +76,20 @@ func (k Keeper) OnConnectChannel(
 	env := types.NewEnv(ctx, contractAddr)
 	querier := k.newQueryHandler(ctx, contractAddr)
 
+	msg := wasmvmtypes.IBCChannelConnectMsg{}
+	if counterpartyVersion == "" {
```
Comment (uint) [inline x/wasm/keeper/relay.go:80]: I don't really know this codebase, but would it be possible to rely on something more meaningful for this?
```go
if msgConfirm {
    // ..
}
```

**Comment by uint** [comment]
LGTM outside of my one comment!

File: x/wasm/keeper/relay.go (hunk 2 above)
Comment (ethanfrey) [reply x/wasm/keeper/relay.go:80]: Good question, captured in https://github.com/CosmWasm/wasmd/issues/561

**Pull request merged**